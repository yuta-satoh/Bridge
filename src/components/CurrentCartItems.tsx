import Image from 'next/image';
import useSWR, { Fetcher } from 'swr';
import myStyles from '../styles/mypage.module.css';
import cciStyles from '../styles/currentCartItems.module.css'

// 型定義
type Items = {
  id: number;
  name: string;
  description: string;
  genre: string;
  category: string;
  price: number;
  imgurl: string;
  stock: number;
  delete: boolean;
};

type Carts = {
  id: number;
  user_id: number;
  delete: boolean;
};

type CartItemsData = {
  id: number;
  item_id: number;
  cart_id: number;
  date: Date;
  quantity: number;
  delete: boolean;
  items: Items;
  carts: Carts;
};

const fetcher: Fetcher<CartItemsData[], string> = (args: string) =>
  fetch(args).then((res) => res.json());

export default function CurrentCartItems({
  cookie,
}: {
  cookie: string;
}) {
  //   SWRでアイテムを取得
  const { data: cartItemData, error } = useSWR<
    CartItemsData[],
    Error
  >(`/api/getCart/items?id=${cookie}`, fetcher);

  // エラー文
  if (error) {
    return (
      <>
        <div className={cciStyles.container}>
          <div className={cciStyles.background}>
            <p>カートに商品がありません</p>
          </div>
        </div>
      </>
    );
  }

  if (!cartItemData) {
    return (
      <>
        <div className="container">
          <div className="background">
            <p>ロード中...</p>
          </div>
        </div>
      </>
    );
  }

  if (cartItemData.length === 0) {
    return (
      <>
        <div className="container">
          <div className="background">
            <p>カートに商品がありません</p>
          </div>
        </div>
      </>
    );
  }

  // 削除フラグがtrueは非表示
  const filteredItemData = cartItemData.filter(
    (item) => !item.delete
  );

  // 上位4つまで抽出(array.slice(0, 3))
  const currentCartItems = filteredItemData.slice(0, 4);

  const otherLength =
    filteredItemData.length - currentCartItems.length;

  return (
    <>
      <div className={cciStyles.cartItemsArea}>
        {currentCartItems.map((cartItem) => {
          return (
            <div className={cciStyles.cartItems} key={cartItem.id}>
              <Image
                src={cartItem.items.imgurl}
                alt={'kagu'}
                width={150}
                height={150}
                className={myStyles.picture}
              />
              <p className={myStyles.cartName}>
                {cartItem.items.name}
              </p>
              <div className={myStyles.cartPrice}>
                <p>
                  ¥ {(cartItem.items.price * 1.1).toLocaleString()}
                </p>
                <p>{`${cartItem.quantity}個`}</p>
              </div>
            </div>
          );
        })}
        {otherLength === 0 ? (
          <></>
        ) : (
          <p className={myStyles.other}>他{otherLength}件</p>
        )}
      </div>
    </>
  );
}
