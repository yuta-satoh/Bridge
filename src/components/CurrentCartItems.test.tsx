import renderer from 'react-test-renderer';
import CurrentCartItems from './CurrentCartItems';
import useSWR, { SWRConfig, SWRResponse } from 'swr';

describe('CarrentCartItemsモジュールのテスト', () => {
  it('ロード中', () => {
    // useSWRがLoadingの場合の定義
    // 戻り値にuseSWRのResponceを返す関数を定義
    const testLoadingMiddleware = () => {
      return (): SWRResponse<any, any> => {
        return {
          data: undefined,
          error: undefined,
          mutate: (_) => Promise.resolve(),
          isValidating: true,
          isLoading: true,
        };
      };
    };

    // 仮のcookieデータを持つpropsを定義
    const props = { cookie: testCookie };
    // CarrentCartItemsにpropsを渡す
    const CarrentCartItems_comp = <CurrentCartItems {...props} />;
    // 描画テスト
    // 「ロード中...」が表示されるはず
    const tree = renderer
      .create(
        // SWRConfigをvalue={{use: [*定義したuseSWRResponceを返す関数*]}}とすると、
        // 疑似的にuseSWRの動きを再現できるらしい？
        <SWRConfig value={{ use: [testLoadingMiddleware] }}>
          {CarrentCartItems_comp}
        </SWRConfig>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('エラー', () => {
    // useSWRのresponceがerrorの場合の定義
    const testErrorMiddleware = () => {
      return (): SWRResponse<any, any> => {
        return {
          data: undefined,
          error: Error(),
          mutate: (_) => Promise.resolve(),
          isValidating: false,
          isLoading: false,
        };
      };
    };

    // 以下、描画テストの手続き(同上)
    const props = { cookie: testCookie };
    const CarrentCartItems_comp = <CurrentCartItems {...props} />;
    const tree = renderer
      .create(
        <SWRConfig value={{ use: [testErrorMiddleware] }}>
          {CarrentCartItems_comp}
        </SWRConfig>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('カートに商品がありません(data.length === 0)', () => {
    // useSWRのdataが空の場合の定義
    const testEmptyMiddleware = () => {
      return (): SWRResponse<any, any> => {
        return {
          data: [],
          error: Error(),
          mutate: (_) => Promise.resolve(),
          isValidating: false,
          isLoading: false,
        };
      };
    };

    // 以下、描画テストの手続き(同上)
    const props = { cookie: testCookie };
    const CarrentCartItems_comp = <CurrentCartItems {...props} />;
    const tree = renderer
      .create(
        <SWRConfig value={{ use: [testEmptyMiddleware] }}>
          {CarrentCartItems_comp}
        </SWRConfig>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('最新の商品4つが表示される', () => {
    // useSWRのdataがある場合の定義
    const testMiddleware = () => {
      return (): SWRResponse<any, any> => {
        return {
          data: test_cart_items,
          error: undefined,
          mutate: (_) => Promise.resolve(),
          isValidating: false,
          isLoading: false,
        };
      };
    };
    // 以下、描画テストの手続き(同上)
    const props = { cookie: testCookie };
    const CarrentCartItems_comp = <CurrentCartItems {...props} />;
    const tree = renderer
      .create(
        <SWRConfig value={{ use: [testMiddleware] }}>
          {CarrentCartItems_comp}
        </SWRConfig>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

// テスト用クッキー
const testCookie: string = '1';
// テスト用cart_itemsデータ
const test_cart_items = [
  {
    id: 1,
    item_id: 1,
    cart_id: 1,
    date: '2023-03-14',
    quantity: 1,
    delete: false,
    items: {
      id: 1,
      name: 'ダミー1',
      price: 10000,
      imgurl: '/hoge1',
    },
  },
  {
    id: 2,
    cart_id: 1,
    date: '2023-03-14',
    quantity: 2,
    delete: false,
    items: {
      id: 2,
      name: 'ダミー2',
      price: 20000,
      imgurl: '/hoge2',
    },
  },
  {
    id: 3,
    cart_id: 1,
    date: '2023-03-14',
    quantity: 3,
    delete: false,
    items: {
      id: 3,
      name: 'ダミー3',
      price: 30000,
      imgurl: '/hoge3',
    },
  },
  {
    id: 4,
    cart_id: 1,
    date: '2023-03-14',
    quantity: 4,
    delete: false,
    items: {
      id: 4,
      name: 'ダミー4',
      price: 40000,
      imgurl: '/hoge4',
    },
  },
  {
    id: 5,
    cart_id: 1,
    date: '2023-03-14',
    quantity: 5,
    delete: false,
    items: {
      id: 5,
      name: 'ダミー5',
      price: 50000,
      imgurl: '/hoge5',
    },
  },
];
