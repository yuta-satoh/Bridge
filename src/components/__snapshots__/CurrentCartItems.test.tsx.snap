// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarrentCartItemsモジュールのテスト エラー 1`] = `
<div
  className="container"
>
  <div
    className="background"
  >
    <p>
      カートに商品がありません
    </p>
  </div>
</div>
`;

exports[`CarrentCartItemsモジュールのテスト カートに商品がありません(data.length === 0) 1`] = `
<div
  className="container"
>
  <div
    className="background"
  >
    <p>
      カートに商品がありません
    </p>
  </div>
</div>
`;

exports[`CarrentCartItemsモジュールのテスト ロード中 1`] = `
<div
  className="container"
>
  <div
    className="background"
  >
    <p>
      ロード中...
    </p>
  </div>
</div>
`;

exports[`CarrentCartItemsモジュールのテスト 最新の商品4つが表示される 1`] = `
<div
  className="cartItemsArea"
>
  <div
    className="cartItems"
  >
    <img
      alt="kagu"
      className="picture"
      data-nimg="1"
      decoding="async"
      height={150}
      loading="lazy"
      onError={[Function]}
      onLoad={[Function]}
      src="/_next/image?url=%2Fhoge1&w=384&q=75"
      srcSet="/_next/image?url=%2Fhoge1&w=256&q=75 1x, /_next/image?url=%2Fhoge1&w=384&q=75 2x"
      style={
        {
          "color": "transparent",
        }
      }
      width={150}
    />
    <p
      className="cartName"
    >
      ダミー1
    </p>
    <div
      className="cartPrice"
    >
      <p>
        ¥ 
        11,000
      </p>
      <p>
        1個
      </p>
    </div>
  </div>
  <div
    className="cartItems"
  >
    <img
      alt="kagu"
      className="picture"
      data-nimg="1"
      decoding="async"
      height={150}
      loading="lazy"
      onError={[Function]}
      onLoad={[Function]}
      src="/_next/image?url=%2Fhoge2&w=384&q=75"
      srcSet="/_next/image?url=%2Fhoge2&w=256&q=75 1x, /_next/image?url=%2Fhoge2&w=384&q=75 2x"
      style={
        {
          "color": "transparent",
        }
      }
      width={150}
    />
    <p
      className="cartName"
    >
      ダミー2
    </p>
    <div
      className="cartPrice"
    >
      <p>
        ¥ 
        22,000
      </p>
      <p>
        2個
      </p>
    </div>
  </div>
  <div
    className="cartItems"
  >
    <img
      alt="kagu"
      className="picture"
      data-nimg="1"
      decoding="async"
      height={150}
      loading="lazy"
      onError={[Function]}
      onLoad={[Function]}
      src="/_next/image?url=%2Fhoge3&w=384&q=75"
      srcSet="/_next/image?url=%2Fhoge3&w=256&q=75 1x, /_next/image?url=%2Fhoge3&w=384&q=75 2x"
      style={
        {
          "color": "transparent",
        }
      }
      width={150}
    />
    <p
      className="cartName"
    >
      ダミー3
    </p>
    <div
      className="cartPrice"
    >
      <p>
        ¥ 
        33,000
      </p>
      <p>
        3個
      </p>
    </div>
  </div>
  <div
    className="cartItems"
  >
    <img
      alt="kagu"
      className="picture"
      data-nimg="1"
      decoding="async"
      height={150}
      loading="lazy"
      onError={[Function]}
      onLoad={[Function]}
      src="/_next/image?url=%2Fhoge4&w=384&q=75"
      srcSet="/_next/image?url=%2Fhoge4&w=256&q=75 1x, /_next/image?url=%2Fhoge4&w=384&q=75 2x"
      style={
        {
          "color": "transparent",
        }
      }
      width={150}
    />
    <p
      className="cartName"
    >
      ダミー4
    </p>
    <div
      className="cartPrice"
    >
      <p>
        ¥ 
        44,000
      </p>
      <p>
        4個
      </p>
    </div>
  </div>
  <p
    className="other"
  >
    他
    1
    件
  </p>
</div>
`;
